pkg/controller/servingcert/controller/secret_updating_controller.go

```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
 
 func (sc *serviceServingCertUpdateController) Sync(obj metav1.Object) error {
 	sharedSecret := obj.(*v1.Secret)
+	service := sc.getServiceForSecret(sharedSecret)
+	if service == nil {
+		return nil
+	}
 
-	regenerate, service := sc.requiresRegeneration(sharedSecret)
-	if !regenerate {
+	if !isSecretValidForService(service, sharedSecret) {
 		return nil
 	}
 
 	// make a copy to avoid mutating cache state
 	secretCopy := sharedSecret.DeepCopy()
 
-	if err := toRequiredSecret(sc.dnsSuffix, sc.ca, service, secretCopy); err != nil {
-		return err
+	update := false
+	if sc.requiresRegeneration(service, sharedSecret, sc.minTimeLeftForCert) {
+		// Regenerate the secret
+		if err := toRequiredSecret(sc.dnsSuffix, sc.ca, service, secretCopy); err != nil {
+			return err
+		}
+		update = true
+	} else {
+		replace := false
+		var err error
+		// If not regenerating, perform checks here to
+		// 1. ensure only the 2 data keys we want exist, TLSCertKey and TLSPrivateKeyKey
+		// 2. ensure cert data is at least a parseable certificate, if not replace with valid data
+		// This does not ensure that somebody didn't swap out secret data for another valid cert.
+		update, replace = sc.ensureSecretData(service, secretCopy)
+		if err != nil {
+			return err
+		}
+		if replace {
```

Comment:
This should have stayed inside `ensureSecretData()`. Its presence there should also explain your question in https://github.com/openshift/service-ca-operator/pull/26/files#r270982624 (`toRequiredSecret()` generates valid data).

Reply:
right, but if we didn't call toRequiredSecret, we should check the cert is a cert. Is toRequiredSecret ok separate from ensureSecretData? It makes it convenient to unit-test, to see whether we are correctly replacing the data or not.